src/llvm-propagate-addrspaces.cpp

```diff
@@ -0,0 +1,233 @@
+        LiftingMap[GEP] = NewGEP;
+        CurrentV = NewGEP;
+    }
+    if (LocTy && cast<PointerType>(CurrentV->getType())->getElementType() != LocTy) {
+        auto *BCI = new BitCastInst(CurrentV, LocTy->getPointerTo());
+        ToInsert.push_back(std::make_pair(BCI, InsertPt));
+        CurrentV = BCI;
+    }
+    return CurrentV;
+}
+
+void PropagateJuliaAddrspaces::visitLoadInst(LoadInst &LI) {
+    unsigned AS = LI.getPointerAddressSpace();
+    if (!isSpecialAS(AS))
+        return;
+    Value *Replacement = LiftPointer(LI.getPointerOperand(), LI.getType(), &LI);
+    if (!Replacement)
+        return;
+    LI.setOperand(LoadInst::getPointerOperandIndex(), Replacement);
+}
+
+void PropagateJuliaAddrspaces::visitStoreInst(StoreInst &SI) {
+    unsigned AS = SI.getPointerAddressSpace();
+    if (!isSpecialAS(AS))
+        return;
+    Value *Replacement = LiftPointer(SI.getPointerOperand(), SI.getValueOperand()->getType(), &SI);
+    if (!Replacement)
+        return;
+    SI.setOperand(StoreInst::getPointerOperandIndex(), Replacement);
+}
+
+void PropagateJuliaAddrspaces::visitMemSetInst(MemSetInst &MI) {
+    unsigned AS = MI.getDestAddressSpace();
+    if (!isSpecialAS(AS))
+        return;
+    Value *Replacement = LiftPointer(MI.getRawDest());
+    if (!Replacement)
+        return;
+    Value *TheFn = Intrinsic::getDeclaration(MI.getModule(), Intrinsic::memset,
+        {Replacement->getType(), MI.getOperand(1)->getType()});
+    CallInst *CI = CallInst::Create(TheFn, {Replacement, MI.getOperand(1)}, "");
+    CI->takeName(&MI);
+#if JL_LLVM_VERSION >= 50000
+    CI->copyMetadata(MI);
+#else
+    SmallVector<std::pair<unsigned, MDNode *>, 4> TheMDs;
+    MI.getAllMetadataOtherThanDebugLoc(TheMDs);
+    for (const auto &MD : TheMDs)
+        CI->setMetadata(MD.first, MD.second);
+    CI->setDebugLoc(MI.getDebugLoc());
+#endif
+ToInsert.push_back(std::make_pair(CI, &MI));
+    ToDelete.push_back(&MI);
+}
+
+void PropagateJuliaAddrspaces::visitMemTransferInst(MemTransferInst &MTI) {
+    unsigned DestAS = MTI.getDestAddressSpace();
+    unsigned SrcAS = MTI.getSourceAddressSpace();
+    if (!isSpecialAS(DestAS) && !isSpecialAS(SrcAS))
+        return;
+    Value *Dest = MTI.getRawDest();
+    if (isSpecialAS(DestAS)) {
+        Value *Replacement = LiftPointer(Dest, cast<PointerType>(Dest->getType())->getElementType(), &MTI);
+        if (Replacement)
+            Dest = Replacement;
+    }
+    Value *Src = MTI.getRawSource();
+    if (isSpecialAS(SrcAS)) {
+        Value *Replacement = LiftPointer(Src, cast<PointerType>(Src->getType())->getElementType(), &MTI);
+        if (Replacement)
+            Src = Replacement;
+    }
+    if (Dest == MTI.getRawDest() && Src == MTI.getRawSource())
+        return;
+    Value *TheFn = Intrinsic::getDeclaration(MTI.getModule(), MTI.getIntrinsicID(),
+        {Dest->getType(), Src->getType(),
+         MTI.getOperand(2)->getType()});
+    CallInst *CI = CallInst::Create(TheFn, {Dest, Src,
```

Comment:
I think it'll be better to just do `MTI->setCalledFunction(TheFn); setArgOperand(0, Dest); setArgOperand(1, Src);`. That way, there's fewer things that need to be moved over to get wrong / outdated.

http://llvm.org/docs/doxygen/html/classllvm_1_1CallInst.html#a61f747edd1ca001427e02e02a320b709